src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,830 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2015-2024 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#include <migraphx/shape_transform_descriptor.hpp>
+#include <migraphx/permutation.hpp>
+#include <migraphx/operation.hpp>
+#include <migraphx/algorithm.hpp>
+#include <migraphx/output_iterator.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/common_dims.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/stringutils.hpp>
+#include <map>
+#include <deque>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+using dimension = shape_transform_descriptor::dimension;
+
+template <class Iterator, class Projection>
+static auto compute_end_dim(Iterator start, Iterator last, std::size_t dim, Projection proj)
+{
+    std::size_t x = 1;
+    auto it       = std::find_if(start, last, [&](auto d) {
+        x *= proj(d);
+        return x >= dim;
+    });
+    if(x != dim)
+        return start;
+    return it;
```

Comment:
```suggestion
    std::size_t x = 1;
    auto it       = std::find_if(start, last, [&](auto d) {
        x *= proj(d);
        return x == dim;
    });
    if(it != last) 
            return it;
    return start;
```

Reply:
No that wont work when the reshape is not divisible by the dimensions. So a reshape from `[2, 3, 5]` to `[10, 3]` wont have an end dim in this case. So we return false, since this would require fractional axes which the shape_transform_descriptor doesnt support. 

Let me add a unit test for this.

Reply:
I do not see why proposed change will not work. Logically it would also return false for the case you mentioned  `[2, 3, 5]` to `[10, 3]`

Reply:
I see, if its last you return `start`. This is different from the original which could return `last` from `find_if`, but looking how its used this would be an error because we do `it + 1` to get the last iterator. I am not sure how we would hit such a case since we check the number of elements are the same.